### ISSUE TITLE
`wflow_publish()` gives `Error in stringr::str_split(raw_git, "\\s")[[1]] :`

### ISSUE THREAD

**Issue by smped** [workflowr/workflowr#188]
Hi.

Thanks for putting all this together. Looks like I'm the only one with this problem though. My project was an existing one with existing commits, but which I added the `workflowr` infrastructure over the top of. Running `wflow_build()` is fine as it just builds any changed files, but I get errors running:

```
wflow_build(update = TRUE)
```
`Error in stringr::str_split(raw_git, "\\s")[[1]] : 
  subscript out of bounds`

Similarly:
```
wflow_publish()
```
`Error in stringr::str_split(raw_git, "\\s")[[1]] : 
  subscript out of bounds`

So it looks like these are the same errors. I've had a dig but can't spot the subroutine that's going awry here. TBH, I'm just trying to get rid of those `R Markdown file: uncommitted changes` warnings. Turns out if I don't commit the Rmd before running `wflow_build()`, I literally can't get rid of those warnings ever, unless I add dummy spaces or something then recommit & rebuild. That seems like a poor solution really.

```
traceback()
4: last_commit_time(repo, files[i], sysgit = sysgit)
3: get_outdated_files(r, absolute(files_analysis[published]), outdir = absolute(o$docs))
2: wflow_status(project = project)
1: wflow_publish()
```

```
sessionInfo()
R version 3.6.2 (2019-12-12)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C               LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8     LC_MONETARY=en_AU.UTF-8   
 [6] LC_MESSAGES=en_AU.UTF-8    LC_PAPER=en_AU.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.6.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.3       rstudioapi_0.10  knitr_1.27       magrittr_1.5     tidyselect_0.2.5 here_0.1         R6_2.4.1         rlang_0.4.2     
 [9] stringr_1.4.0    dplyr_0.8.3      tools_3.6.2      xfun_0.12        git2r_0.26.1     htmltools_0.4.0  yaml_2.2.0       digest_0.6.23   
[17] rprojroot_1.3-2  assertthat_0.2.1 tibble_2.1.3     crayon_1.3.4     processx_3.4.1   callr_3.4.0      purrr_0.3.3      later_1.0.0     
[25] ps_1.3.0         promises_1.1.0   fs_1.3.1         glue_1.3.1       evaluate_0.14    rmarkdown_2.0    stringi_1.4.5    pander_0.6.3    
[33] pillar_1.4.3     compiler_3.6.2   backports_1.1.5  httpuv_1.5.2     pkgconfig_2.0.3 
```

**Comment by jdblischak** [comment]
@steveped Thanks for the bug report!

Sorry for the inconvenience. This is a new error, but it is due to the behavior of your system Git executable, similar to the errors in #180 and #182.

For a quick fix, you can force workflowr to use the git2r package instead of your machine's system Git executable by adding the following to the `.Rprofile` in your project's root directory:

```
options(workflowr.sysgit = "")
```

Then try running:

```
wflow_publish(".Rprofile", update = TRUE)
```

For context, workflowr tries to use the system Git executable because it is much faster. However, it is more error-prone to try and parse the output of Git, which is what caused the error in this case. Using git2r, which calls libgit2 (a minimal C implementation of Git), is much safer.

> I literally can't get rid of those warnings ever, unless I add dummy spaces or something then recommit & rebuild. That seems like a poor solution really.

Once this bug is fixed, you won't need to do anything like that. `wflow_publish()` is designed to handle all of that. It will commit the Rmd files if needed, so that warning will be removed.

To help me troubleshoot this issue and prevent other users from having the same issue, could you please run the following in the terminal and report the results?

```
which git
git --version
git log -n 1 --date=raw --format=%ad -- analysis/index.Rmd
```

**Comment by smped** [comment]
Hi @jdblischak ,
Thanks for the speedy response! I'll give that a run and will try figure my way around `git2r`. As to your suggested information, I get:
```
$ which git
/usr/bin/git
$ git --version
git version 2.17.1
$ git log -n 1 --date=raw --format=%ad -- analysis/index.Rmd
1579507914 +1030
```
Hopefully that's vaguely informative.
Cheers

**Comment by jdblischak** [comment]
> will try figure my way around git2r

No need to do that! (unless you want to of course). I was just providing you some background information on the internals of workflowr. However, you certainly do not need to know anything about git2r to use workflowr. One of the main goals of workflowr is to make it easy to get started using version control without first having to spend hours going through tutorials.

> Hopefully that's vaguely informative.

That was exactly the information I wanted. But it's puzzling. Your setup is essentially identical to mine (same R, Git, and Ubuntu):

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.3 LTS
Release:	18.04
Codename:	bionic
$ Rscript --version
R scripting front-end version 3.6.2 (2019-12-12)
$ which git
/usr/bin/git
$ git --version
git version 2.17.1
$  git log -n 1 --date=raw --format=%ad -- analysis/index.Rmd
1559224774 -0400
```

So some more follow-up questions:

1. Are there any spaces in the path to your workflowr project? For example, is there a space in your Ubuntu login name or the name of your workflowr project?
1. Are you able to share your workflowr project so I can test it?
1. What do you get when you create a temporary example workflowr project using the code below?
```
cwd <- getwd()
setwd(tempdir())
library(workflowr)
rmd <- tempfile(fileext = ".Rmd")
file.create(rmd)
wflow_quickstart(rmd, "github-username", create_on_github = FALSE)
wflow_status()
setwd(cwd)
```

**Comment by smped** [comment]
Thanks John. Always love a bug that's hard to reproduce. In answer to your questions:

1. No. Spaces in file paths are the work of the devil :)
3. Everything worked fine in your example, even running `wflow_publish()` and `wflow_build(update = TRUE)`.

As to sharing the project, there's a bit of unpublished stuff there my collaborators may baulk at sharing. I'd be happy to extract any git logs, or those type of files though as I suspect it's parsing an edge case from those that may be my problem. Let me know if that's workable at all.

**Comment by jdblischak** [comment]
> No. Spaces in file paths are the work of the devil :)

Ha! Glad we're on the same page.

> Everything worked fine in your example, even running wflow_publish() and wflow_build(update = TRUE).

That's promising. This means that there is an isolated issue specific to your project that is causing the problem.

> As to sharing the project, there's a bit of unpublished stuff there my collaborators may baulk at sharing. I'd be happy to extract any git logs, or those type of files though as I suspect it's parsing an edge case from those that may be my problem. Let me know if that's workable at all.

No worries. I can have you investigate:

1. Did setting `options(workflowr.sysgit = "")` fix your problem? Or was there a different error?
1. Did you run `wflow_start()` with `existing = TRUE` to introduce the the workflowr infrastructure to your existing project? Did you customize any of the workflowr settings?
1. Could you please run the diagnostic commands below and report the output?

```
workflowr:::wflow_paths()
rmarkdown::yaml_front_matter("analysis/index.Rmd")
file.exists("_workflowr.yml")
file.exists("analysis/_site.yml")
yaml::read_yaml("analysis/_site.yml")$output
yaml::read_yaml("analysis/_site.yml")$output_dir
```

Thanks for your help troubleshooting. Much appreciated!

**Comment by jdblischak** [comment]
@steveped Any update on the troubleshooting? Please let me know if there is anything I can do to help.

**Comment by smped** [comment]
Hi @jdblischak ,

Thanks for chasing & my apologies for the slow response. Beaten by juggling multiple projects & meeting deadlines this week. Firstly, the error has simply disappeared, without any adjustment to my `.Rprofile`. I fear it's the 'have you tried turning it off & on again' syndrome, however I had actually tried that before posting the issue.

All my arguments appear to be the defaults in your diagnostic code too. I have deleted the license file from the repo & YAML though as I didn't feel it to be a necessary component for this particular project.

```
> workflowr:::wflow_paths()
$root
[1] "."

$analysis
[1] "analysis"

$docs
[1] "docs"

$git
[1] "."

> rmarkdown::yaml_front_matter("analysis/index.Rmd")
$title
[1] "Home"

$site
[1] "workflowr::wflow_site"

$output
$output$`workflowr::wflow_html`
$output$`workflowr::wflow_html`$toc
[1] FALSE



$editor_options
$editor_options$chunk_output_type
[1] "console"


> file.exists("_workflowr.yml")
[1] TRUE
> file.exists("analysis/_site.yml")
[1] TRUE
> yaml::read_yaml("analysis/_site.yml")$output
$`workflowr::wflow_html`
$`workflowr::wflow_html`$toc
[1] TRUE

$`workflowr::wflow_html`$toc_float
[1] TRUE

$`workflowr::wflow_html`$theme
[1] "cosmo"

$`workflowr::wflow_html`$highlight
[1] "textmate"

$`workflowr::wflow_html`$code_folding
[1] "hide"


> yaml::read_yaml("analysis/_site.yml")$output_dir
[1] "../docs"
```

Thanks again for your engagement here too. It's great to see a project that's genuinely supported. As I'm sure you're aware, orphaned software is a huge problem in bioinformatics.

**Comment by jdblischak** [comment]
> Firstly, the error has simply disappeared, without any adjustment to my .Rprofile. I fear it's the 'have you tried turning it off & on again' syndrome, however I had actually tried that before posting the issue.

Strange. Though I'm glad it's working now! Please do let me know if you ever find it behaving strangely again. Like I said above, `wflow_publish()` is designed to make it easy to commit new changes without you have to manually `git add` anything.

> I have deleted the license file from the repo & YAML though as I didn't feel it to be a necessary component for this particular project.

That's no problem. Most files are simply recommendations, and even the required parts can be renamed (e.g. you can change the name of `analysis/` to whatever you like). The only real requirements are 1) a Git repo somewhere, 2) an `.Rproj` file at the root of the project, and 3) the files `_site.yml` and `index.Rmd` saved in the same directory.

However, there is one consequence of deleting `_workflowr.yml`. By default it sets `knit_root_dir: "."`, which causes the code in the R Markdown chunks to be executed from the root of the project. If this variable isn't set,then workflowr defaults to the default knitr behavior of running the code in the same directory as the Rmd file.

> Thanks again for your engagement here too. It's great to see a project that's genuinely supported.

Of course! And I appreciate you taking the time to file a detailed bug report.

>  As I'm sure you're aware, orphaned software is a huge problem in bioinformatics.

I agree it is a frustrating problem that I have run into many times myself. Though it is important to remember that this is more the fault of the scientific funding agencies than any individual scientist. It is incredibly difficult to get funding for software maintenance, especially "infrastruture" software like workflowr. We are lucky that our lab received a grant from a private foundation, the [Gordon and Betty Moore Foundation][moore], which allows us to develop and maintain workflowr and other open source scientific software.

[moore]: https://www.moore.org/

**Comment by rgayler** [comment]
I have just run into this problem for the first time:
```
> wflow_status()
Error in stringr::str_split(raw_git, "\\s")[[1]] : 
  subscript out of bounds
```
I tried the diagnostics you suggested above:
```
$ which git
/usr/bin/git
$ git --version
git version 2.30.2
$git log -n 1 --date=raw --format=%ad -- analysis/index.Rmd
1623920433 +1000
```
I created a new git branch from the RStudio IDE and committed my changes, then `wflow_status()` no longer errored.
I switched back to the main branch - no error.

So the problem appears to have gone away. I am only mentioning it here in case it suggests something useful to @jdblischak .

**Comment by jdblischak** [comment]
> So the problem appears to have gone away. I am only mentioning it here in case it suggests something useful

Thanks @rgayler! Please report back if it starts to regularly happen. Your setup looks fine, so this shouldn't be happening.

**Comment by smped** [comment]
Thanks for the tip @rgayler. I've had this crop up a couple of times randomly, but I just commit manually & it seems to fix itself. Tried the changing branches trick & that seemed to work.

**Comment by FloWuenne** [comment]
Hi @jdblischak ,

I have run into a similar error today. I was working on a project I initialized using workflowR and that had been working in the past. Today I got the error described above. I then added the following to .Rprofile as suggested by you:
```
options(workflowr.sysgit = "")
```
This did unfortunately not fix the error, but changed it to the following: 
```
Error in git2r::commits(repo, n = 1, path = fname)[[1]] : 
  subscript out of bounds
```
I don't really know how to fix this at this point... I had to recently play around with https vs git pushing for another repo and maybe I changed some settings in git that could be related here? The remote setup in this repo is using https though, which I think is correct... 

Here are some of the diagnostics you had asked other people for:

```
$ which git
/usr/bin/git

$ git --version
git version 2.39.2 (Apple Git-143)

$git log -n 1 --date=raw --format=%ad -- analysis/index.Rmd
1686573093 +0200
```

Any help is appreciated ☺️, thanks!

**Comment by jdblischak** [comment]
@FloWuenne Thanks for the detailed bug report. You've already tried `options(workflowr.sysgit = "")`, and I don't see anything obviously wrong with your setup.

The error in `git2r:commits()` is concerning. It's good that workflowr is using git2r (ie it is correctly interpreting `workflowr.sysgit`), but this suggests that no commits have been made to the file being interrogated. What command did you run to produce that error (`wflow_status()`, `wflow_publish()`)?

https://github.com/workflowr/workflowr/blob/89ba735746f53521ae9f782034790552a1e25690/R/git.R#L223-L227

In the meantime, here is some diagnostic code to run in your workflowr project. Please report the output:

```R
system("git status")
library("git2r")
repo <- repository(".")
status(repo)
branches(repo, flags = "local")
commits(repo, n = 1)
commits(repo, n = 1, path = "README.md")
sessionInfo()
```

**Comment by FloWuenne** [comment]
Hi @jdblischak ,

so both `wflow_status()` and `wflow_publish()` create the same error.

Brace yourself, here the long diagnostics:

![image](https://github.com/workflowr/workflowr/assets/11639154/516d8299-b989-491e-bd2b-df173a82b7b8)

![image](https://github.com/workflowr/workflowr/assets/11639154/25d6c912-84b0-4d1b-93c0-d7cdd29bb948)

![image](https://github.com/workflowr/workflowr/assets/11639154/efa4d337-bd6b-48fe-bcfa-2bd91990b632)

![image](https://github.com/workflowr/workflowr/assets/11639154/f3484da5-f884-437e-a061-db0ef169dfae)

![image](https://github.com/workflowr/workflowr/assets/11639154/85f8002c-ada0-4579-92ae-6e5b2f47c5ea)

![image](https://github.com/workflowr/workflowr/assets/11639154/0a238f80-8306-4269-82ea-8e22e2efc42c)

**Comment by jdblischak** [comment]
Great, thanks for that. Maybe a specific Rmd file is causing the problem. Please try the following next. Which commands succeed and which fail?

```R
wflow_status("analysis/index.Rmd")
wflow_status("analysis/figures.figure_5.Rmd")
wflow_status("analysis/figures.supplementary_figure_5.Rmd")
rmd <- Sys.glob("analysis/*Rmd")
n_rmd <- length(rmd)
wflow_status(rmd[seq(1, n_rmd/2)])
wflow_status(rmd[seq(n_rmd/2 + 1, n_rmd)])
```

**Comment by FloWuenne** [comment]
Thanks @jdblischak !

So to not list too much stuff here first, everything passes until this:

`wflow_status(rmd[seq(1, n_rmd/2)])` -> Same error as before.

![grafik](https://github.com/workflowr/workflowr/assets/11639154/494cb331-fbe4-45e4-8ec3-ab266ccf1b07)

n_rmd is 15.

`wflow_status(rmd[seq(n_rmd/2 + 1, n_rmd)])`  -> Passes without error. Output below:

![grafik](https://github.com/workflowr/workflowr/assets/11639154/6cc0c058-e568-4c1b-97b7-882125bfe763)

**Comment by jdblischak** [comment]
Great. That confirms that not all the Rmd files are problematic. Below is some code to brute-force determine which Rmd files produce the git2r error:

```sh
rmd <- Sys.glob("analysis/*Rmd")
for (i in seq_along(rmd)) {
  attempt <- try(wflow_status(rmd[i]))
  if (inherits(attempt, "try-error")) {
    writeLines(rmd[i])
  }
}
```

**Comment by FloWuenne** [comment]
Thanks so much @jdblischak , that identified the issue. I am still not sure why that particular file is causing issues. For now, I moved it out of the analysis folder and published everything else. Will gradually rebuild the file and see when it breaks. The error message from your command was the following:

```
Error in git2r::commits(repo, n = 1, path = fname)[[1]] : 
  subscript out of bounds
analysis/figures.supplementary_figure_3.Rmd
```

**Comment by jdblischak** [comment]
> that identified the issue

Great! I'm glad we found the problem file and you were able to work around it

> I am still not sure why that particular file is causing issues

I'd love to know why this one file was the problem. If I knew how the situation arose, I could try to prevent it from happening in the first place, or at minimum provide a better error message that explains how to fix it.

My guess is that maybe the HTML and/or figure files were committed, but not the original Rmd. Could you please try running the following Git commands in the terminal?

```sh
git log analysis/figures.supplementary_figure_3.Rmd
git log docs/figures.supplementary_figure_3.html
git log docs/figure/figures.supplementary_figure_3/
```

**Comment by FloWuenne** [comment]
So I tried to run the `git log` commands, but because I did move the faulty file and seem to have commited and pushed since then, the log commands don't actually yield anything (even when pointing to the current location of the files...). I will have a look into why this happened later at some point. For now I actually have to finish this project.

I will try to update the cause of this failure here once I figure it out. To me, it came a bit out of the blue and was really hard to troubleshoot. 

Thank you so much for fixing the issue for me @jdblischak 🙏!
